grakn-graql/src/main/java/ai/grakn/graql/internal/query/QueryOperationExecutor.java

```diff
@@ -424,35 +428,35 @@ GraknTx tx() {
      */
     @AutoValue
     static abstract class VarAndProperty {
+
         abstract Var var();
         abstract VarPropertyInternal property();
 
         static VarAndProperty of(Var var, VarProperty property) {
-            return new AutoValue_InsertQueryExecutor_VarAndProperty(var, VarPropertyInternal.from(property));
+            return new AutoValue_QueryOperationExecutor_VarAndProperty(var, VarPropertyInternal.from(property));
         }
 
         static Stream<VarAndProperty> fromPattern(VarPatternAdmin pattern) {
             return pattern.getProperties().map(prop -> VarAndProperty.of(pattern.var(), prop));
         }
 
-        void insert(InsertQueryExecutor executor) {
-            property().insert(var(), executor);
-        }
-
-        void define(InsertQueryExecutor executor) {
-            property().define(var(), executor);
-        }
-
-        Set<Var> requiredVars() {
-            return property().requiredVars(var());
-        }
-
-        Set<Var> producedVars() {
-            return property().producedVars(var());
+        private PropertyExecutor executor(ExecutionType executionType) {
+            switch (executionType) {
```

Comment:
Ughhhh . . . . I hate switch statements with the obligatory "Cannot reach this line because java". 

maybe it would be better to turn `ExecutionType` into a more complex enum, one which provides the relevant property rather than relying on a switch.

I am thinking you make execution type be:

```
private enum ExecutionType {
    INSERT((var) -> property().insert(var)), 
    DEFINE((var) -> property(). define(var))

    private Function<Var, PropertyExecutor > function;

    public ExecutionType(Function<Var, PropertyExecutor > function){
        this.function = function;
    }

    public property(Var var){
        return function.apply(var);
    }
}
```
This is A LOT MORE VEROSE but it kills the ugly switch and it makes sure that if you add another `ExecutionType` you are forced to define its `PropertyExecutor`